lib/auth/auth.go

```diff
@@ -4478,6 +4495,72 @@ func (a *Server) DeleteNamespace(namespace string) error {
 	return a.Services.DeleteNamespace(namespace)
 }
 
+// ListAccessRequests is an access request getter with pagination and sorting options.
+func (a *Server) ListAccessRequests(ctx context.Context, req *proto.ListAccessRequestsRequest) (*proto.ListAccessRequestsResponse, error) {
+	// most access request methods target the backend directly since access requests are frequently read
+	// immediately after writing, but listing requires support for custom sort orders so we route it to
+	// a special cache. note that the access request cache will still end up forwarding single-request
+	// reads to the real backend due to the read after write issue.
+	return a.AccessRequestCache.ListAccessRequests(ctx, req)
+}
+
+func (a *Server) IterateAccessRequests(ctx context.Context, req *proto.ListAccessRequestsRequest, match func(*types.AccessRequestV3) (bool, error)) ([]*types.AccessRequestV3, string, error) {
+	const maxIterations = 100_000
+
+	if req.Limit == 0 {
+		req.Limit = apidefaults.DefaultChunkSize
+	}
+
+	req.Limit++
+	defer func() {
+		req.Limit--
+	}()
+
+	var filtered []*types.AccessRequestV3
+	var iterations int
+Outer:
+	for {
+		iterations++
+		if iterations > maxIterations {
+			return nil, "", trace.Errorf("too many access request page iterations (%d), this is likely a bug", iterations)
+		}
+
+		rsp, err := a.ListAccessRequests(ctx, req)
+		if err != nil {
+			return nil, "", trace.Wrap(err)
+		}
+
+	Inner:
+		for _, accessRequest := range rsp.AccessRequests {
+			ok, err := match(accessRequest)
+			if err != nil {
+				return nil, "", trace.Wrap(err)
+			}
+			if !ok {
+				continue Inner
+			}
+			filtered = append(filtered, accessRequest)
+			if len(filtered) == int(req.Limit) {
+				break Outer
+			}
+		}
+
+		if rsp.NextKey == "" {
+			break Outer
+		}
+
+		req.StartKey = rsp.NextKey
+	}
+
+	var nextKey string
+	if len(filtered) == int(req.Limit) {
+		nextKey = filtered[req.Limit-1].GetName()
+		filtered = filtered[:req.Limit-1]
+	}
```

Comment:
Good catch, always using ID as the pagination key was a mistake that slipped through because most of the sorting tests didn't use remote clients.

Updated tests to cover custom sorting orders via remote clients and refactored how RBAC filtering works so that we always end up using the pagination key generated by the cache.